File: frontend/src/api/notifications.ts
import { apiClient } from './client';

export interface NotificationResponse {
  id: number;
  user_id: number;
  actor_id: number | null;
  actor_name: string | null;
  title: string;
  message: string | null;
  type: string;
  related_entity_type: string | null;
  related_entity_id: number | null;
  is_read: boolean;
  created_at: string;
}

export const notificationsApi = {
  list: async (unreadOnly: boolean = false): Promise<NotificationResponse[]> => {
    const response = await apiClient.get<NotificationResponse[]>(
      `/notifications?unread_only=${unreadOnly}&size=50`
    );
    return response.data;
  },

  unreadCount: async (): Promise<number> => {
    const response = await apiClient.get<{ count: number }>('/notifications/unread-count');
    return response.data.count;
  },

  markRead: async (notificationId: number): Promise<void> => {
    await apiClient.post(`/notifications/${notificationId}/read`);
  },

  markAllRead: async (): Promise<void> => {
    await apiClient.post('/notifications/mark-all-read');
  },
};

File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session, joinedload
from app.core.database import get_db
from app.core.auth import get_current_user
from app.models import User, Notification
from app.schemas import NotificationResponse
from typing import List

router = APIRouter(prefix="/notifications", tags=["Notifications"])


def _notif_to_response(n: Notification) -> NotificationResponse:
    return NotificationResponse(
        id=n.id,
        user_id=n.user_id,
        actor_id=n.actor_id,
        actor_name=n.actor.full_name if n.actor else None,
        title=n.title,
        message=n.message,
        type=n.type,
        related_entity_type=n.related_entity_type,
        related_entity_id=n.related_entity_id,
        is_read=n.is_read,
        created_at=n.created_at,
    )


@router.get("", response_model=List[NotificationResponse])
def list_notifications(
    unread_only: bool = Query(False),
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List notifications for current user."""
    query = db.query(Notification).options(
        joinedload(Notification.actor)
    ).filter(
        Notification.user_id == current_user.id
    )
    
    if unread_only:
        query = query.filter(Notification.is_read == False)
    
    query = query.order_by(Notification.created_at.desc())
    
    skip = (page - 1) * size
    notifs = query.offset(skip).limit(size).all()
    return [_notif_to_response(n) for n in notifs]


@router.get("/unread-count")
def unread_notification_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get count of unread notifications."""
    count = db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False
    ).count()
    return {"count": count}


@router.post("/{notification_id}/read", status_code=status.HTTP_204_NO_CONTENT)
def mark_notification_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Mark a notification as read."""
    notification = db.query(Notification).filter(
        Notification.id == notification_id,
        Notification.user_id == current_user.id
    ).first()
    
    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )
    
    notification.is_read = True
    db.commit()


@router.post("/mark-all-read", status_code=status.HTTP_204_NO_CONTENT)
def mark_all_notifications_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Mark all notifications as read."""
    db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False
    ).update({"is_read": True})
    db.commit()
